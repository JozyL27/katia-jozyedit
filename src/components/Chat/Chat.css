.chat__main {
  height: 100vh;
  display: flex;
  flex-direction: column;
  justify-content: space-between;
  position: relative;
}

.chat__nav {
  height: 13vmin;
  max-height: 11rem;
  background-color: white;
  display: flex;
  justify-content: center;
  align-items: center;
  box-shadow: 0 1rem .8rem #888888;
  border: none;
  box-shadow: 0 .8rem .8rem #888888;
}

.chat__back {
  position: fixed;
  top: 1.8vh;
  left: 1.8vw;
  width: auto;
  height: 7vmin;
  max-height: 6rem;
}

.chat__chatting-with {
  font-size: 6.5vmin;
  font-weight: bold;
  max-width: 70vmin;
  white-space: nowrap;
  overflow: hidden;
  text-overflow: ellipsis;
  color: black;
}

.chat__message-container {
  display: flex;
  flex-direction: column;
  justify-content: flex-start;
  height: 100%;
  min-height: calc(100vh - 28vmin);
  overflow-y: auto;
  padding: 2vh 3vw 0 3vw;
}

.chat__message-container > :first-child {
  margin-top: auto;
}

.chat__none {
  height: 100%;
  display: flex;
  justify-content: center;
  align-items: center;
}

.chat__none-message {
  width: 75%;
  font-size: 3vmax;
  text-align: center;
}

.chat__message {
  display: flex;
  align-items: flex-end;
  min-height: 8vmin;
  margin: 1vmin;
  width: 100%;
}

.chat__user-message {
  justify-content: flex-end;
}

.chat__message-text {
  position: relative;
  z-index: -1;
  font-size: 3.5vmin;
  padding: 2vmin 1vmin;
  border-radius: 1vmin;
  margin: 0 1vmin;
  max-width: 87vw;
}

.chat__user-message-text {
  background-color: rgb(161, 211, 228);
}

.chat__partner-message-text {
  background-color: rgb(80, 194, 80);
}

.chat__avatar {
  position: relative;
  z-index: -1;
  width: auto;
  height: 8vmin;
}

.chat__message-form {
  background-color: white;
  height: 15vmin;
  width: 100vw;
  display: flex;
  justify-content: space-around;
  align-items: center;
  padding: 0 1vmin;
}

.chat__message-form > * {
  font-size: 5vmin;
}

.chat__input {
  width: 85vw;
  height: 10vmin;
  resize: none;
  padding: .5vmin;
  border-radius: 1vmin;
}

.chat__send {
  width: 9vmin;
  height: auto;
}

.chat__send:hover {
  cursor: pointer;
}

@media screen and (min-width: 900px) {
  .chat__chatting-with {
    font-size: 30px;
  }

  .chat__none-message {
    font-size: 20px;
  }

  .chat__back {
    height: 40px;
    margin-top: 5px;
  }
}