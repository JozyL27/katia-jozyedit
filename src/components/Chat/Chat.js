import React, { Component } from 'react';
import { Link } from 'react-router-dom'
import UserContext from '../../Contexts/UserContext';
import send_icon from '../../images/send-icon.svg';
import backArrow from '../../images/left-arrow-svgrepo-com.svg'
import './Chat.css';

class Chat extends Component {
  state = {
    message: '',
  }
  
  static contextType = UserContext;

  componentDidMount() {
    this.scrollToBottom()
  }

  componentDidUpdate() {
    this.scrollToBottom()
  }

  scrollToBottom = () => {
    this.messagesEnd.scrollIntoView({ behavior: "smooth" });
  }

  handleSendMessage = (event) => {
    event.preventDefault();
    this.props.onSend(this.state.message);
    this.setState({message: ''})
  }

  handleMessageChange = (event) => {
    this.setState({ message: event.target.value });
  }

  generateUserMessage = (user, message) => {
    return (<li key={message.id} className='chat__message chat__user-message' aria-live='polite'>
      <span className='chat__message-text chat__user-message-text'>{message.message}</span>
      <img className='chat__avatar chat__user-avatar' src={user.avatar} alt='user-avatar' />
    </li>)
  }

  generatePartnerMessage = (partner, message) => {
    return (<li key={message.id} className='chat__message chat__partner-message' aria-live='polite'>
      <img className='chat__avatar chat__partner-avatar' src={partner.avatar} alt='partner-avatar' />
      <span className='chat__message-text chat__partner-message-text'>{message.message}</span>
    </li>)
  }

  render() {
    const { messages, user_id, user, partner } = this.props;
    return ( 
      <main className='chat__main'>
        <nav className='chat__nav'>
          <Link to='/matches'>
            <img className='chat__back' src={backArrow} alt='back-arrow' />
          </Link>
          <span className='chat__chatting-with'>{partner.display_name}</span>
        </nav>
        <ul className='chat__message-container'>
          {messages.length === 0 ? <li className='chat__none'><p className='chat__none-message'>Start the conversation by sending the first message</p></li>
              : 
            messages.map(message => {
              return message.sender_id === user_id ? this.generateUserMessage(user, message) : this.generatePartnerMessage(partner, message)
          })}
          <div style={{ float:"left", clear: "both" }}
               ref={(el) => { this.messagesEnd = el; }}>
          </div>
        </ul>
        <form onSubmit={this.handleSendMessage} className='chat__message-form' style={{backgroundColor: this.context.backgroundColor}}>
          <textarea className='chat__input' onChange={this.handleMessageChange} onKeyDown={event => {if (event.keyCode === 13) return this.handleSendMessage(event)}} value={this.state.message}/>
          <img className='chat__send' onClick={this.handleSendMessage} src={send_icon} alt='send'/>
        </form>
      </main>
    );
  }
}
 
export default Chat;