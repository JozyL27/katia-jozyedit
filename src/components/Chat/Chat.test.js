import React from 'react'
import ReactDOM from 'react-dom'
import renderer from 'react-test-renderer'
import { MemoryRouter } from 'react-router-dom'
import Chat from './Chat'
import { UserProvider } from '../../Contexts/UserContext'

describe('<Chat />', () => {
  const testData = {
    messages: [
      {
        id: 1,
        conversation_id: 1,
        sender_id: 1,
        message: 'hi'
      },
      {
        id: 2,
        conversation_id: 1,
        sender_id: 2,
        message: 'sup'
      }
    ],
    user_id: 1,
    user: {
      display_name: 'Test User 1',
      avatar: 'https://katia-app.s3-us-west-1.amazonaws.com/default_avatar.png'
    },
    partner: {
      display_name: 'Test User 2',
      avatar: 'https://katia-app.s3-us-west-1.amazonaws.com/default_avatar.png'
    }
  }
  // Smoke test
  it('renders without crashing', () => {

    window.HTMLElement.prototype.scrollIntoView = function() {};

    const div = document.createElement('div')
    ReactDOM.render(
      <MemoryRouter>
        <UserProvider>
          <Chat messages={testData.messages} user_id={testData.user_id} user={testData.user} partner={testData.partner} />
        </UserProvider>
      </MemoryRouter>, div)
    ReactDOM.unmountComponentAtNode(div)
  })
  // Snapshot test
  it('renders the UI as expected', () => {
    const tree = renderer
      .create(
        <MemoryRouter>
          <UserProvider>
            <Chat messages={testData.messages} user_id={testData.user_id} user={testData.user} partner={testData.partner} />
          </UserProvider>
        </MemoryRouter>,
        {
          createNodeMock: (element) => {
            if(element.type === 'div') {
              return {
                scrollIntoView: (obj) => {
                  return true;
                }
              };
            }
            return true;
          }
        }
      )
      .toJSON()
    expect(tree).toMatchSnapshot()
  })
})