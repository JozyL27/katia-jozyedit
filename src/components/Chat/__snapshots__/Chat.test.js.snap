// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`<Chat /> renders the UI as expected 1`] = `
<main
  className="chat__main"
>
  <nav
    className="chat__nav"
  >
    <a
      href="/matches"
      onClick={[Function]}
    >
      <img
        alt="back-arrow"
        className="chat__back"
        src="left-arrow-svgrepo-com.svg"
      />
    </a>
    <span
      className="chat__chatting-with"
    >
      Test User 2
    </span>
  </nav>
  <ul
    className="chat__message-container"
  >
    <li
      aria-live="polite"
      className="chat__message chat__user-message"
    >
      <span
        className="chat__message-text chat__user-message-text"
      >
        hi
      </span>
      <img
        alt="user-avatar"
        className="chat__avatar chat__user-avatar"
        src="https://katia-app.s3-us-west-1.amazonaws.com/default_avatar.png"
      />
    </li>
    <li
      aria-live="polite"
      className="chat__message chat__partner-message"
    >
      <img
        alt="partner-avatar"
        className="chat__avatar chat__partner-avatar"
        src="https://katia-app.s3-us-west-1.amazonaws.com/default_avatar.png"
      />
      <span
        className="chat__message-text chat__partner-message-text"
      >
        sup
      </span>
    </li>
    <div
      style={
        Object {
          "clear": "both",
          "float": "left",
        }
      }
    />
  </ul>
  <form
    className="chat__message-form"
    onSubmit={[Function]}
    style={
      Object {
        "backgroundColor": undefined,
      }
    }
  >
    <textarea
      className="chat__input"
      onChange={[Function]}
      onKeyDown={[Function]}
      value=""
    />
    <img
      alt="send"
      className="chat__send"
      onClick={[Function]}
      src="send-icon.svg"
    />
  </form>
</main>
`;
